tools/RunGen.cpp

```diff
@@ -0,0 +1,1019 @@
+            break;
+        }
+        case halide_argument_kind_output_buffer:
+            // Nothing yet
+            break;
+        }
+    }
+
+    // Run a bounds query, so we can allocate output buffers appropriately.
+    {
+        for (auto &arg_pair : args) {
+            auto &arg = arg_pair.second;
+            switch (arg.metadata->kind) {
+            case halide_argument_kind_output_buffer:
+                auto bounds_query_shape = choose_output_extents(arg.metadata->dimensions, default_output_shape);
+                arg.buffer_value = Buffer<>(arg.metadata->type, nullptr, (int) bounds_query_shape.size(), &bounds_query_shape[0]);
+                filter_argv[arg.index] = arg.buffer_value.raw_buffer();
+                break;
+            }
+        }
+
+        info() << "Running bounds query...";
+        int result = halide_rungen_redirect_argv(&filter_argv[0]);
+        if (result != 0) {
+            fail() << "Bounds query failed with result code: " << result;
+        }
+    }
+
+    // Allocate the output buffers we'll need.
+    double pixels_out = 0.f;
+    for (auto &arg_pair : args) {
+        auto &arg_name = arg_pair.first;
+        auto &arg = arg_pair.second;
+        switch (arg.metadata->kind) {
+        case halide_argument_kind_output_buffer:
+            auto constrained_shape = get_shape(arg.buffer_value);
+            info() << "Output " << arg_name << ": BoundsQuery result is " << constrained_shape;
+            Shape shape = fix_bounds_query_shape(constrained_shape);
+            arg.buffer_value = allocate_buffer(arg.metadata->type, shape);
+            info() << "Output " << arg_name << ": Shape is " << get_shape(arg.buffer_value);
+            filter_argv[arg.index] = arg.buffer_value.raw_buffer();
+            // TODO: this assumes that most output is "pixel-ish", and counting the size of the first
+            // two dimensions approximates the "pixel size". This is not, in general, a valid assumption,
+            // but is a useful metric for benchmarking.
+            if (shape.size() >= 2) {
+                pixels_out += shape[0].extent * shape[1].extent;
+            } else {
+                pixels_out += shape[0].extent;
+            }
+            break;
+        }
+    }
+    double megapixels = pixels_out / (1024.0 * 1024.0);
+
+    // If we're tracking memory, install the memory tracker *after* doing a bounds query.
+    HalideMemoryTracker tracker;
+    if (track_memory) {
+        tracker.install();
+    }
+
+    if (benchmark) {
+        info() << "Benchmarking filter...";
+
+        // Run once to warm up cache. Ignore result since our halide_error() should catch everything.
+        (void) halide_rungen_redirect_argv(&filter_argv[0]);
+
+        double time_in_seconds = Halide::Tools::benchmark(benchmark_samples, benchmark_iterations, [&filter_argv]() { 
+            (void) halide_rungen_redirect_argv(&filter_argv[0]);
```

Comment:
Add output.device_sync(). Otherwise we're just measuring how long it takes to do a kernel launch when we're using the gpu